Improve tooltips for torrent progress bar

evsh opened the pull request:
A little improvements for the progress bars in the "General" tab of the torrent properties are proposed.

1. Show files (and their sizes) for a piece under cursor in tooltip.
2. If the cursor points to a single file which extends over several pieces, highlight those pieces.

This works as follows: the following phrase is added to the tooltip "Hold Shift key for detailed information":
Here is an illustration:
![qbt-new-progress-bar-simple-crop](https://cloud.githubusercontent.com/assets/394621/13990206/c1cfb09a-f113-11e5-9a53-02664f92743d.png)
And if this modifier (Shift) is pressed when tooltip is requested, detailed information about a pieces of a file under cursor is shown:
![qbt-new-progress-detailed-crop](https://cloud.githubusercontent.com/assets/394621/13990280/fda29f10-f113-11e5-8409-b4816ccbd863.png)
The detailed tooltip shows a table with sizes and names of the files in the piece under cursor. 

In both images one can see highlighted (actually darkened) part of the torrent, occupied by the file under cursor.

This PR extends (and based on) PR #4867. The work continues in PR #4994

LordNyriox commented:
@evsh:  You should probably go ahead and close the other peer-list PR (use this one instead).  It doesn't make much sense without the additional commits here anyway.  


Thanks, 
~ Nyriox

LordNyriox commented:
@evsh:  Also, you might want to check your PR for Qt 4 compatibility.  All the Travis builds for Qt 4 have failed (though the Qt 5 builds worked fine).  

Thanks, 
~ Nyriox

sledgehammer999 commented:
>You should probably close [#4867] and use this one instead.

No! Don't do that.

evsh commented:
@LordNyriox , thank you! I always forget to replace those Q_DECL_OVERRIDE inserted by Qt Creator...

LordNyriox commented:
@evsh:  Glad to be of help.

Danny3 commented:
@evsh 
Since you're working on this...
Could you maybe add the the actual progress bar that exists in μTorrent but it's missing here?
It's 4 pixels tall surrounded by a 1 pixel pixel grey border as far as I can tell.
![progress_bar_cropped](https://cloud.githubusercontent.com/assets/1429543/13723479/0d9c2db0-e867-11e5-9ae2-e365516a6e88.png)
This exist in the older version of μTorrent (1.8.5) that I'm using.
This matches the percentage and can be seen how it's moving when downloading a torrent fast or when the torrent is rechecked.

It's a little unexpected for me that the progress bar doesn't really show the progress of downloading torrent, but the downloaded parts of it and the real progress bar is just the small one shown int the Done column.

evsh commented:
@Danny3 to me this "progress line" looks like a separate widget. Why does the progress percentage not satisfy you? Do you want just a graphical representation or there is something more in this progress bar?

LordNyriox commented:
@evsh:  The point of a progress bar is typically to allow an at-a-glance view of current progress.  

The piece preview is typically too scattered to allow that at-a-glance understanding, even though the average user would by default open up that tab and expect an ordinary progress bar as well. 

Personally, I'd like to see this feature as well (just like it was described above, if possible), but it would probably be best to implement it in a separate PR, so as not to make this one too complicated.  

Thanks, 
~ Nyriox

LordNyriox commented:
@evsh:  Once @sledgehammer999 merges your peer-list PR, you'll need to rebase this one.

LordNyriox commented:
@evsh:  PR [https://github.com/qbittorrent/qBittorrent/pull/4867] is now merged.  You should go ahead and rebase this (removing the first 2 commits while you are at it).  

Thanks, 
~ Nyriox

Danny3 commented:
@evsh

I found the "progress line" to be useful in 2 cases:
**1**. A little trick that I found myself:
When I'm taking a break, like going o eat or something, I just leave the pointer of my mouse at the end of it.
When I get back, I look how much the end of line has moved from where my mouse pointer is.
If it didn't move, it's clear to me that it didn't download anything more.
I don't think I'll remember the percentange, especially if it's a floating point number and this method seems faster than writing the number in a text file for later comparison.
**2**. Good to detect slowdowns on hard drives.
If i remember right, I think that once I detected that the antivirus was running a scan because I was looking at the progress line after doing a force rechek on a torrent and it seemed to me that the line was moving too slow compared to other times.
I think that I wouldn't have noticed anything looking at the percentage number.
Maybe it's my brain that likes graphs more than numbers...
And it's easier for it to see a grahical overview than the exact number.

I thought that  qBittorrent tried to offer a simmilar UI to μTorrent keeping as much as possible of the good features in μTorrent.
This seems to me like a good feature which takes little vertical space.
If yo can't add it, don't wory, it's ok.
Thanks for asking why do I need it for.
BTW, about this PR, showing the files in the tooltip looks like a nice feature.

evsh commented:
@Danny3, thank you for the explanations! Thinking about this progress bar, may I ask just one more question? Why don't you use the ETA column in the torrent list? Is it because the column relies on instantaneous speed? What do you think about a small addition to this progress line: to show completed in blue and active (donwloading) in green if the width of this addition is big enough (> 1px?)?

LordNyriox commented:
@evsh:  The coloring certainly sounds great to me.  :}  

ETA column is better for sorting torrents by finish date than it is for an at-a-glance view of overall progress within one torrent.  I sort by the ETA column (and can remember numbers), but even I have wanted a simple straight progress bar sometimes.  

Also, I think 5-10 pixels high *should* be enough for this progress bar, but I leave that decision to you.  :} 

Thanks, 
~ Nyriox

Chocobo1 commented:
Just tried it out, cool function!

Here's some bugs I observed:
* I start qbt with empty jobs, the bars looks like it's floating, but it should be flat as before, see pics below:
Before:
![before_hover](https://cloud.githubusercontent.com/assets/9395168/13751642/144591e8-ea46-11e5-9407-bf0ffc275a19.png)
After I hover mouse over Progress bar, it becomes normal:
![after_hover](https://cloud.githubusercontent.com/assets/9395168/13751643/1445e4c2-ea46-11e5-9d8e-fae13372aff0.png)
This floating bug is also observed when I have only 1 jobs and I removed it, then the bar start to be floating.

* About the tooltips text, maybe it's better to switch the legends and pieces info and use the string "This piece contains:" as title.
```
White: .....
(and other legends)

This piece contains:
```

* in tooltips, use only filename, leave out it's path.

What do you think?

evsh commented:
@Chocobo1,thank you for the comments! I will definitely investigate the first problem, because the strange thing is I changed the bar drawing code to avoid the same "hovered" impression on my system. Perhaps this is just because of the difference between light and dark backgrounds. 
I think that the path in the tooltip is useful. Think about a torrent with series of something. Than each subdirectory may contain files with names like 001, 002, 003, .. Without path information they all will be indistinguishable.

evsh commented:
I implemented this feature for the following situation. Suppose I want to download a book. I launch a search and (see PR #3989) obtain a number of book collections where this particular book is present. Then I start to download a torrent and find that it is incomplete with current peers. But do they have the book I need? Now I can overview the torrent and check. If no, I can try another one.
And here is the question: perhaps I should instead add an "Availability" column to the "Contents" tab? This column might show instantaneous availability for the torrent files.

Chocobo1 commented:
>Without path information they all will be indistinguishable.

yeah, then it's better to have path included.
But I just felt too much text crammed together (not sure how to describe it precisely).

Another idea, is it better to move file size to the start of a line? that way I can quickly know what file (the piece contains) by simply looking from the end (of a string).

Danny3 commented:
@evsh 
Most of the times the ETA column shows ∞ (Infinity symbol)
The tracker could be down or the only one seeder could be offline at that moment.
IMO, not showing anything it would be the same, either way, it doesn't help me at all.
When it works it's just a smart guess based on the current speed.
Unfortunately it has a major problem that's impossible to fix.
It assumes that the speed will remain constant, (tracker will stay online and peers will remain online and upload at the same speed).
Anyway, it gives you an idea by predicting the **future**, it's good enough, it's nice.
I will rephrase my example with taking a break from the computer:
Let's say I add a tutorial torrent in the evening.
ETA or no ETA, this will take a while and I put my mouse pointer to the end of the progress line and go to bed.
When I wake up and come to computer I take a look how much progress the torrent made over the night from where I left the mouse.
I don't care about the time, because I know it, it's from when I went to bed until now when I'm looking.
So I don't care that much about the future that ETA shows, but the progress in the **past**, that has already happened and it's accurate.
I think the progress it's useful for multiple days ETA torrents too, if you want to quickly see how much progress the torrent made each night. leaving the mouse there, for curiosity.

Normally I like simple straight progress bar (one color), but your idea seems interesting and good for showing more detailed information without requiring extra space and I will like to give it a try.
I assume making both versions with a settings toggle would be too much work.
BTW, are you sure that having the second color "green" wouldn't be a problem for color blind people like the ones that have Deuteranopia (problem seeing green)?
http://www.color-blindness.com/deuteranopia-red-green-color-blindness/
Maybe they will see only the first part (blue) and not the the second (green), so they will see half a line...

Chocobo1 commented:
@evsh 
One last thing for this round :p
were you using tabs for indentation? IMO it took too much (precious) spaces, maybe just 2~3 spaces is enough?

evsh commented:
No, this default indentation from HTML <ul> tag.

evsh commented:
Can I mark somehow the PR as not yet ready for merge? There is a bug in it: if I select a torrent with downloaded metadata and after that a torrent without it, the progress bar shows the progress of the former one after mouse moved out of it.

LordNyriox commented:
@evsh:  From what I've seen with other PRs you just make a comment explicitly stating it's not ready to merge (like you just did).  

@Chocobo1:  Do you have any idea what might be causing this bug?  

Thanks, 
~ Nyriox

LordNyriox commented:
@evsh:  Also, I do like the 'Availability' column idea (and would like to see it implemented as well), but I feel that having a graphical option (like this PR) might be more attractive to the end user.  

Do you think you can open a separate PR for the `Availability` column (which might be merged sooner), but keep working on this one as well?  (Best of both worlds:  graphical and simple viewing options.)  

Thanks, 
~ Nyriox

Chocobo1 commented:
>Can I mark somehow the PR as not yet ready for merge? 

just add something like `[don't merge]` in the title or in the first post

LordNyriox commented:
@Chocobo1:  While you're still around, do you think you could look into what's causing @evsh the bug he mentioned?  

See comment [https://github.com/qbittorrent/qBittorrent/pull/4936#issuecomment-199084955]. 

Thanks, 
~ Nyriox

evsh commented:
Thank you all for the suggestion! I've worked around the bug by cleaning piece information in `clean()`. Perhaps it is better to do that anyway.

evsh commented:
Fixed drawing error, now with light theme it should be correct. Tried to remove indentation from the list, but without success (can someone help me, please? How to change indentation in `<ul>` element rendered in tooltip?). Tried to typeset the files information in a table. The code is there, under TYPESET_FILELIST_AS_TABLE define. However, I got strange alignment with hieroglyphs in file names: file names were shifted downwards for approx 0.1-0.2 ex. Maybe somebody can tell what's wrong there? Moved file size information to the beginning of the list items.

LordNyriox commented:
@evsh:  Thanks for all your hard work!  

In reply to [https://github.com/qbittorrent/qBittorrent/pull/4936#issuecomment-196417251]:  (Yes, I know I'm repeating myself a bit, but as usual I commented while half asleep.  Silly me.)  I do think adding an `Availability` column to the `Content` tab is a good idea (and something I think you should do in a separate PR), but this graphical representation is more important right now.  

I think you should do both, in short.  ;} 

Thanks, 
~ Nyriox

ngosang commented:
I'd prefer the Legend at the top of the tooltip.

evsh commented:
I decided to put the legend at the bottom because otherwise the tooltip text looks like a filled with water balloon, pinned by mouse pointer, which is not aesthetic. Could you, maybe, suggest how to avoid that look? Which spaces do you want to remove?

ngosang commented:
> Could you, maybe, suggest how to avoid that look? 

No, but most relevant/important info should be first. For users it is more important to understand the graphic.

> Which spaces do you want to remove?

Before:
```
Files in these pieces:

   *  Image....

White: ...
Green: ...
Blue: ...
```

After:
```
White: ...
Green: ...
Blue: ...
Files in these pieces:
- Image....
```

evsh commented:
@ngosang, @Chocobo1  What do you think about the following: we can show the "simple" tooltip (only legend), but if one of the modifier keys is pressed, we show another tooltip which contains _only_ the file list.  Additionally, we can hide file size with another modifier (or this is too much?)
So, the simple tooltip contains the following:
White: ...
Green: ...
Blue: ...
Hold Shift for detailed information

And the detailed tooltip does not contain the legend at all. This saves us from ugly text pyramid and from annoying spaces, but leaves the tooltips clear and readable.

Chocobo1 commented:
@evsh 
Good idea!

> we can show the "simple" tooltip (only legend), but if one of the modifier keys is pressed, we show another tooltip which contains only the file list. 

agree.

>we can hide file size with another modifier (or this is too much?)

too cumbersome.

better add parentheses: `(Hold Shift for detailed information)`

evsh commented:
Implemented the tooltips split. Activated table rendering by default for the detailed tooltip, because I think that vertically aligned cells look good. I did not add the suggested parentheses because, as to me, now the simple tooltip is clean and nice.

LordNyriox commented:
@evsh:  Don't forget to update the duplicated commits in PR [https://github.com/qbittorrent/qBittorrent/pull/4994] as well.  

Thanks, 
~ Nyriox

evsh commented:
Detailed tooltip caption is left-aligned now

evsh commented:
I think this is ready for merging.

ngosang commented:
Sometimes I see the size in two lines but I think it should be always in one line. Maybe using a non-breaking space...
![](http://storage5.static.itmages.com/i/16/0507/h_1462621722_3737736_a648343370.png)

evsh commented:
@ngosang: I've added 'white-space:nowrap' to the style of the size table cell. Seem to be working here. Could you test, please? But I agree with you that there has to be the thin unbreakable space between the number and the unit.

sledgehammer999 commented on src/base/bittorrent/torrentinfo.cpp:
```diff
@@ -213,24 +214,79 @@ QByteArray TorrentInfo::metadata() const
+    std::transform(fileIndices.begin(), fileIndices.end(), std::back_inserter(res), [this](int i) {
+        return this->filePath(i);
+    });
+
```
Are we inventing new ways to use every possible feature of the C++ language in qbittorrent? I would do it in a for loop here.
PS: I don't want you to change it. Just a remark.

evsh replied:
Perhaps it's just me, but `std::transform` shows the purpose more clearly because the STL algorithms are declarative.

sledgehammer999 replied:
Another point is that templates usually bloat the the file size. (again not saying you should change it)

evsh replied:
Should we have 
```C++
template <class T>
A complexFunction(T v)
{
...
}
```
and call it like
```C++
int i; short s;
complexFunction(i);
complexFunction(s);
```
the binary size will increase comparing to

```C++
A complexFunction(int v)
{
...
}

int i; short s;
complexFunction(i);
complexFunction(s);
```

But here? Why?

sledgehammer999 replied:
It was a general comment.

evsh replied:
OK

sledgehammer999 commented on src/gui/properties/downloadedpiecesbar.h:
```diff
@@ -32,68 +32,39 @@
-    // buffered 256 levels gradient from bg_color to piece_color
-    QVector<int> m_pieceColors;
+    void setProgress(const QBitArray &m_pieces, const QBitArray &downloadedPieces);
 
```
delete newline

sledgehammer999 commented on src/gui/properties/downloadedpiecesbar.h:
```diff
@@ -32,68 +32,39 @@
+    // TODO: make a diff pieces to new pieces and update only changed pixels, speedup when update > 20x faster
+    QBitArray m_pieces;
+    QBitArray m_downloadedPieces;
 
```
delete newline

sledgehammer999 commented on src/gui/properties/downloadedpiecesbar.h:
```diff
@@ -32,68 +32,39 @@
-
-class DownloadedPiecesBar: public QWidget {
     Q_OBJECT
-    Q_DISABLE_COPY(DownloadedPiecesBar)
```
Repeat the macro because of https://stackoverflow.com/questions/19854371/repeating-q-disable-copy-in-qobject-derived-classes

sledgehammer999 commented on src/gui/properties/piecesbar.h:
```diff
@@ -0,0 +1,107 @@
+    void setColors(const QColor &background, const QColor &border, const QColor &complete);
+
+    // QObject interface
+    virtual bool event(QEvent *) override;
```
I think it is better to organize it this way:
```c++
public:
    explicit PiecesBar(QWidget *parent = nullptr);

    void setTorrent(BitTorrent::TorrentHandle *torrent);
    void setColors(const QColor &background, const QColor &border, const QColor &complete);

    virtual void clear();
    // QObject interface
    virtual bool event(QEvent *) override;
```

sledgehammer999 commented on src/gui/properties/piecesbar.cpp:
```diff
@@ -0,0 +1,325 @@
+
+        int imagePos(int pieceIndex) const
+        {
+            return static_cast<float>(pieceIndex) / (m_piecesCount - 1) * (m_imageWidth - 1);
```
potential division by zero if `m_piecesCount == 1` or `m_imageWidth == 1`

evsh replied:
Well, the code is certainly not clean here and relies on the fact that when we cast float infinity to int, we receive `-MAX_INT` and this value is discarded in the next checks.

evsh replied:
But I will fix the code, of course.

sledgehammer999 commented on src/gui/properties/piecesbar.cpp:
```diff
@@ -0,0 +1,325 @@
+
+        int pieceIndex(int imagePos) const
+        {
+            return static_cast<float>(imagePos) / (m_imageWidth - 1) * (m_piecesCount - 1);
```
same here

sledgehammer999 commented on src/gui/properties/piecesbar.cpp:
```diff
@@ -0,0 +1,325 @@
+    , m_pieceColors{}
+    , m_hovered{false}
+    , m_highlitedRegion{}
+{
```
Do you really need to initialize explicitly `m_image`, `m_pieceColors` and `m_highlitedRegion`?

evsh replied:
Like to be explicit.

sledgehammer999 replied:
Lets leverage the language and not be very verbose with no reason. If it has a default constructor don't empty initialize it. So pelase delete those 3.

evsh replied:
OK

sledgehammer999 commented on src/gui/properties/piecesbar.cpp:
```diff
@@ -0,0 +1,325 @@
+{
+#define TYPESET_FILELIST_AS_TABLE
+
+    class DetailedTooltipRenderer
```
I think this should go to the top like the other class.

sledgehammer999 commented on src/gui/properties/piecesbar.cpp:
```diff
@@ -0,0 +1,325 @@
+    class DetailedTooltipRenderer
+    {
+    public:
+        DetailedTooltipRenderer(QTextStream& stream, const QString &header)
```
`QTextStream &stream`

sledgehammer999 commented on src/gui/properties/piecesbar.cpp:
```diff
@@ -0,0 +1,325 @@
+            QList<int> files {m_torrent->info().fileIndicesForPiece(pieceIndex)};
+
+            DetailedTooltipRenderer renderer(stream,
+                                             files.count() > 1 ? tr("Files in this piece:") : tr("Files in these pieces:"));
```
I think you have to wording wrong here. It should be:
`files.count() > 1 ? tr("File in this piece:") : tr("Files in this piece:"));`

evsh replied:
We highlight several pieces if a file is stretched over them.

evsh replied:
And only a single piece if it contains more than one file.

sledgehammer999 replied:
But here you test "files.count()" which doesn't tell you if the file spans multiple pieces.

evsh replied:
This is the number of files in the piece under the mouse pointer. If the files size is not exactly equal to the piece size, this works.

sledgehammer999 replied:
Shit I was reading the comparison wrong. But it still can lead to weird outputs. For example one file that occupies one whole piece. The tooltip will read "Files in these pieces" which is wrong.

evsh replied:
Then I add a check for the files size if `files.count() == 1`

sledgehammer999 replied:
Should I add more pedantry?
Three strings:
1. "File in this piece" -> one file that occupies only one whole piece
2. "Files in these pieces" -> more than one file that occupy one piece
3. "File in these pieces" -> one file that occupies multiple pieces

What happens if user hovers over a piece that has 2 files one of which spans to the next(or previous) piece? Do you highlight those pieces too?

evsh replied:
> Three strings:

OK

> What happens if user hovers over a piece that has 2 files one of which spans to the next(or previous) piece? Do you highlight those pieces too?

No, I don't. Only the piece under the cursor is highlighted.

sledgehammer999 commented on src/gui/properties/piecesbar.cpp:
```diff
@@ -0,0 +1,325 @@
+
+bool PiecesBar::isFileNameCorrectionNeeded() const
+{
+    return false;
```
What is the purpose of this if you're always going to return the same value? For exposing it in the future via a setting?

evsh replied:
The function is virtual, and the overrided one in `PieceAvailabilityBar` returns `true`.

sledgehammer999 commented on src/gui/properties/piecesbar.cpp:
```diff
@@ -0,0 +1,325 @@
+            const bool isFileNameCorrectionNeeded = this->isFileNameCorrectionNeeded();
+            for (int f: files) {
+                QString filePath {m_torrent->info().filePath(f)};
+                if (!isFileNameCorrectionNeeded)
```
Shouldn't you drop the exclamation mark here?

evsh replied:
I must have made a mistake in `isFileNameCorrectionNeeded()` because the generated tooltip texts are correct.

evsh replied:
Or not :)

evsh replied:
We do not want to se ".unwanted" in the `PiecesAvailiabilityBar` (remote files), right?

sledgehammer999 replied:
No we don't. But the name of the function indicates that you don't need the exclamation mark. So you need to changed the returned values or the function name to avoid confusions.

evsh replied:
Then the negation has to be removed.

sledgehammer999 replied:
Now that I am thinking it, is there really a situation where we want to output the ".unwanted" folder? I think not.

evsh replied:
I think for the "Progress" bar (local files) we should output exactly what is on the drive, i.e. including ".unwanted".

sledgehammer999 replied:
Yeah. You're right. I am sleepy. I can't think straight.

sledgehammer999 commented on src/gui/properties/pieceavailabilitybar.h:
```diff
@@ -31,52 +31,32 @@
+class PieceAvailabilityBar: public PiecesBar
 {
     Q_OBJECT
-    Q_DISABLE_COPY(PieceAvailabilityBar)
```
for the same reason I think you should repeat this macro here too

sledgehammer999 commented on src/gui/properties/piecesbar.cpp:
```diff
@@ -0,0 +1,325 @@
+
+namespace
+{
+#define TYPESET_FILELIST_AS_TABLE
```
Why this define? I see the different code paths, but why do they achieve?
Is the random user that compiler supposed to know about this define and define/undefine it himself?

evsh replied:
I was not sure which layout to use for tooltips and asked people to compare. You may do this too, but after some recent adjustments to the table layout it is almost perfect and before merging I will delete the define and unused code path.

sledgehammer999 commented:
I think I am done with the code review. I didn't check it architecturally-wise much. I don't see something obvious.
I see that @glassez hasn't comment. Can you take a look too?

PS: The use of `{}` instead of `()` for initializing stuff becomes annoying to read and IMO breaks the style the rest of the repo uses.

evsh commented:
@glassez  did that already.

sledgehammer999 commented:
>@glassez did that already.

Are you sure? I don't see him in the "participants" list at the right-panel of this page.

evsh commented:
> Are you sure? 

He did that in #4994, which includes these changes.

sledgehammer999 commented:
Some design decision which I think is intentional but I am asking generally in case others agree. (I am not pushing this change unless other people agree): The file size in the tooltip is printed in the middle of each filename. Wouldn't it be better to print it at the top?

evsh commented:
On my system and with my files the file names usually occupy two lines. Aligned to the middle size labels allows me to match them easily.

sledgehammer999 commented on src/gui/properties/downloadedpiecesbar.h:
```diff
@@ -37,49 +37,63 @@
+{
+    class TorrentHandle;
+}
+
```
And I was wondering from where these came from(I was looking at the 2nd commit diff only).
These don't belong in this commit.

evsh replied:
Ups...

sledgehammer999 commented on src/gui/properties/downloadedpiecesbar.h:
```diff
@@ -37,49 +37,63 @@
+
+    // QObject interface
+public:
+    virtual bool event(QEvent *) Q_DECL_OVERRIDE;
```
This also doesn't belong in this commit.

sledgehammer999 commented:
Can you run uncrustify on your first commit? I noticed if/elses that don't need braces and some that have their braces in the wrong line.
PS: If it gives too much trouble with the 2nd commit, then I guess you should do it for the downloadedpiecesbar.h/cpp files and merge the changes into the 2nd commit.

glassez commented:
>He did that in #4994, which includes these changes.

I'm not sure I did it in detail. So I'll see it here .

evsh commented:
@sledgehammer999, I believe all your comments have been addressed. Give me a note, please, when to remove `#define TYPESET_FILELIST_AS_TABLE`.

sledgehammer999 commented on src/gui/properties/downloadedpiecesbar.h:
```diff
@@ -32,54 +32,39 @@
 
-    // I used values, because it should be possible to change colors in runtime
+    void setProgress(const QBitArray &m_pieces, const QBitArray &downloadedPieces);
 
```
remove newline

sledgehammer999 commented on src/gui/properties/piecesbar.cpp:
```diff
@@ -0,0 +1,336 @@
+
+        int pieceIndex(int imagePos) const
+        {
+            return m_imageWidth == 1 ? 0
```
you need to check for `m_piecesCount` too. Remember it is multiplication and any zero makes the whole multiplicaiton zero.

evsh replied:
Same here.

sledgehammer999 commented on src/gui/properties/piecesbar.cpp:
```diff
@@ -0,0 +1,336 @@
+
+        int imagePos(int pieceIndex) const
+        {
+            return m_piecesCount == 1 ? 0
```
you need to check for `m_imageWidth` too. Remember it is multiplication and any zero makes the whole multiplicaiton zero.

evsh replied:
Apparently, there is need for brackets here, because I read this expression incorrectly yesterday too. `m_imageWidth` is in nominator.

sledgehammer999 replied:
You're right. Division and multiplication have the same priority and you go from left to right. But better put the parentheses to be sure.

evsh replied:
Wouldn't it be more readable to replace `m_imageWidth` and `m_piecesCount` with `float m_pixelsPerPiece`?

sledgehammer999 replied:
Yes, but why float? You can't have eg half a pixel.

evsh replied:
But we need to store ratio with precision of about 1e-3 (size of a typical display) in order to transform piece index to pixel with 1-pixel precision. If a torrent contains less than 1000 pieces, integer ratio is not enough.

sledgehammer999 replied:
I am probably confused on how you calculate pixelsPerPiece. How do you do it?

evsh replied:
`torrent->info().piecesCount() / image.width()`

sledgehammer999 replied:
Then yes float makes sense.
Remember 2 special cases:
1. The last piece might be smaller than piece_size
2. I can't explain it well. But take into consideration that due to rounding errors the edges (first/last piece) might be off by some pixels. Not too important.

sledgehammer999 commented:
Those were my final comments I think. I tested the code it looks good. So now I am waiting for @glassez

glassez commented:
@evsh `Follow project coding style. Issue #2192.` means "Fix indentation" only again?

evsh commented:
means "run uncrustify"

glassez commented on src/base/bittorrent/torrentinfo.cpp:
```diff
@@ -213,24 +220,81 @@ QByteArray TorrentInfo::metadata() const
+
+    QStringList res;
+    res.reserve(fileIndices.size());
+    std::transform(fileIndices.begin(), fileIndices.end(), std::back_inserter(res), [this](int i) {
```
Fix coding style for lambda. If you need split it you should use same style as for other functions:
```c++
[]()
{
    // body
}
```

glassez replied:
```c++
std::transform(fileIndices.begin(), fileIndices.end(), std::back_inserter(res), [this](int i)
{
    return filePath(i);
});
```

glassez commented on src/base/bittorrent/torrentinfo.cpp:
```diff
@@ -213,24 +220,81 @@ QByteArray TorrentInfo::metadata() const
+    QStringList res;
+    res.reserve(fileIndices.size());
+    std::transform(fileIndices.begin(), fileIndices.end(), std::back_inserter(res), [this](int i) {
+        return this->filePath(i);
```
**this** isn't required here.

glassez commented on src/base/bittorrent/torrentinfo.cpp:
```diff
@@ -213,24 +220,81 @@ QByteArray TorrentInfo::metadata() const
+    return res;
+}
+
+QVector<int> BitTorrent::TorrentInfo::fileIndicesForPiece(int pieceIndex) const
```
Remove `BitTorrent::` prefix from newly added methods.

glassez commented on src/base/bittorrent/torrentinfo.cpp:
```diff
@@ -213,24 +220,81 @@ QByteArray TorrentInfo::metadata() const
-    }
+    QVector<int> res;
+    res.reserve(files.size());
+    std::transform(files.begin(), files.end(), std::back_inserter(res), [](const libtorrent::file_slice &s) {
```
Fix coding style as I described above.

glassez commented on src/base/bittorrent/torrentinfo.cpp:
```diff
@@ -213,24 +220,81 @@ QByteArray TorrentInfo::metadata() const
+
+    int index = fileIndex(file);
+    if (index == -1) {
+        qDebug() << "Filename '" << file << "' was not found in torrent '" << name() << '\'';
```
Minor. Do you remember QDebug stream output adds spaces so you will get `Filename ' file ' was not found in torrent ' torrent name '`.

glassez commented on src/base/bittorrent/torrentinfo.cpp:
```diff
@@ -213,24 +220,81 @@ QByteArray TorrentInfo::metadata() const
+        return qMakePair(-1, -1);
+    }
+
+    const libtorrent::file_storage &files = nativeInfo()->files();
```
You can use `libt` shorthand in BitTorrent \*.cpp files.

glassez commented on src/base/bittorrent/torrentinfo.h:
```diff
@@ -86,12 +89,20 @@ namespace BitTorrent
+
+        // returns pair of the first and the last pieces into which
+        // the given file extends (maybe partially).
+        QPair<int,int> piecesForFile(const QString &file) const;
```
`QPair<int, int>`
Doesn't your uncrustify know this?

evsh replied:
It can not process just two lines.

glassez commented on src/gui/properties/downloadedpiecesbar.cpp:
```diff
@@ -28,28 +28,24 @@
 
-    updatePieceColors();
+DownloadedPiecesBar::DownloadedPiecesBar(QWidget *parent)
+    : base{parent}
```
Fix coding style for initializers.

glassez commented on src/gui/properties/downloadedpiecesbar.h:
```diff
@@ -32,54 +32,39 @@
+    DownloadedPiecesBar(QWidget *parent);
 
-    // I used values, because it should be possible to change colors in runtime
+    void setProgress(const QBitArray &m_pieces, const QBitArray &downloadedPieces);
```
Fix `m_pieces` => `pieces`.

glassez commented on src/gui/properties/pieceavailabilitybar.cpp:
```diff
@@ -28,20 +28,15 @@
 
 PieceAvailabilityBar::PieceAvailabilityBar(QWidget *parent)
-    : QWidget(parent)
+    : base{parent}
```
Fix coding style.

glassez commented on src/gui/properties/pieceavailabilitybar.h:
```diff
@@ -31,52 +31,33 @@
+    void clear() override;
+
+private:
+    virtual bool updateImage(QImage &image) override;
```
*virtual* is redundant here!

sledgehammer999 replied:
I think it is better to leave it. If one is reading he should know immediately that it is virtual.

glassez replied:
non-virtual CANNOT BE override. In other words: override implies virtual.

sledgehammer999 replied:
I missed "override".

glassez commented on src/gui/properties/piecesbar.cpp:
```diff
@@ -0,0 +1,336 @@
+    class PieceIndexToImagePos
+    {
+    public:
+        PieceIndexToImagePos(const BitTorrent::TorrentHandle*const torrent, const QImage &image)
```
Insert space befor `*`.

glassez commented on src/gui/properties/piecesbar.cpp:
```diff
@@ -0,0 +1,336 @@
+}
+
+PiecesBar::PiecesBar(QWidget *parent)
+    : QWidget{parent}
```
Fix coding style for initializers.

glassez replied:
And please don't use uncrustify now unless you sure it will not produce another coding style violations. But you can use discovered violations to upgrade uncrustify.cfg later.

glassez commented on src/gui/properties/piecesbar.cpp:
```diff
@@ -0,0 +1,336 @@
+    return m_pieceColor;
+}
+
+const QVector<QRgb>&PiecesBar::pieceColors() const
```
Insert space before `&`.

glassez replied:
Of course, you can use `QVector<QRgb>` instead of const reference.

glassez commented on src/gui/properties/piecesbar.cpp:
```diff
@@ -0,0 +1,336 @@
+    int g2 = qGreen(rgb2);
+    int b2 = qBlue(rgb2);
+
+    float ratio_n = 1.0 - ratio;
```
Fix `ratio_n` name.

glassez replied:
You forget to fix this issue.

evsh replied:
It's not my code, but OK.

glassez commented on src/gui/properties/piecesbar.cpp:
```diff
@@ -0,0 +1,336 @@
+    bool showDetailedInformation = QApplication::keyboardModifiers().testFlag(Qt::ShiftModifier);
+    if (!showDetailedInformation) {
+        stream << simpleToolTipText();
+        stream << '\n' << tr("Hold Shift key for detailed information");
```
Maybe `return` after to avoid too long `else`?

glassez commented on src/gui/properties/piecesbar.cpp:
```diff
@@ -0,0 +1,336 @@
+                    tooltipTitle = tr("File in these pieces");
+            }
+
+            DetailedTooltipRenderer renderer(stream, tooltipTitle);
```
If you're going to use it in a functional context, it is better to give it a verbal name: render.
But, IMO, I would change its `operator()` to `void render()`.

glassez replied:
And what is your comment about this?

evsh replied:
I do not want to change the names. The object *is* renderer, and `operator()` is its natural "do the only thing you can" method.

glassez replied:
Okay, let it remain as is.

glassez commented on src/gui/properties/piecesbar.cpp:
```diff
@@ -0,0 +1,336 @@
+    int pieceIndex = transform.pieceIndex(imagePos);
+    QVector<int> fileIndices {m_torrent->info().fileIndicesForPiece(pieceIndex)};
+    if (fileIndices.count() == 1) {
+        QPair<int,int> fileBounds = m_torrent->info().piecesForFile(fileIndices.first());
```
Fix `QPair<int,int>` => `QPair<int, int>`

glassez commented on src/gui/properties/piecesbar.h:
```diff
@@ -0,0 +1,109 @@
+    virtual void clear();
+
+    // QObject interface
+    virtual bool event(QEvent*) override;
```
Are you sure it should be public?

evsh replied:
Well, it is public in `QObject`, so there is no big difference...

glassez replied:
Then OK.

glassez commented on src/gui/properties/piecesbar.h:
```diff
@@ -0,0 +1,109 @@
+    void showToolTip(const QHelpEvent*);
+    void highlightFile(int imagePos);
+
+    virtual QString simpleToolTipText() const = 0;
```
private virtual/abstract methods??? Tear out my eyes!

evsh replied:
?

glassez replied:
Why not `protected`?
I don't see much point in making virtual methods private. It only confuses people.

evsh replied:
Because a descendant does not need (and thus disallowed) to call it.

glassez replied:
Okay. Although they are still abstract.

evsh replied:
Herb Sutter once wrote a nice [blog](http://www.gotw.ca/publications/mill18.htm) on this subject.

glassez commented:
>means "run uncrustify"

Then rename commit accordingly, please.

glassez commented:
@evsh I've finished commenting. Please don't force push here some time after you answered to my comments to not complicate our dialogue (so I don't have to look outdated diffs every time).

glassez commented:
@evsh it would be better if you will first add a separate commit with fixes, and then (after I review it) meld it into previous.

evsh commented:
> @evsh it would be better if you will first add a separate commit with fixes, and then (after I review it) meld it into previous.

OK, take a look, please.

glassez commented on src/gui/properties/piecesbar.cpp:
```diff
@@ -255,11 +252,7 @@ void PiecesBar::showToolTip(const QHelpEvent *e)
     bool showDetailedInformation = QApplication::keyboardModifiers().testFlag(Qt::ShiftModifier);
-    if (!showDetailedInformation) {
-        stream << simpleToolTipText();
-        stream << '\n' << tr("Hold Shift key for detailed information");
```
I mean insert `return` here and then drop `else` and put its content one indentation level up.

evsh replied:
Does that mean "create two functions out of this one"?

glassez replied:
Ok. Forget it. Now I'll see there is some code after `if-else`.

glassez commented on src/base/bittorrent/torrentinfo.cpp:
```diff
@@ -225,53 +225,57 @@ QStringList TorrentInfo::filesForPiece(int pieceIndex) const
-        return this->filePath(i);
-    });
+    std::transform(fileIndices.begin(), fileIndices.end(), std::back_inserter(res),
+        [this](int i)
```
If you moved it that way (not the way I showed), you could do it in one line:
```c++
std::transform(fileIndices.begin(), fileIndices.end(), std::back_inserter(res), 
    [this](int i) { return filePath(i); });
```

glassez commented:
@evsh, @sledgehammer999 I've finished reviewing. I have not delved into the rendering code, I hope it works correctly.

evsh commented on src/gui/properties/piecesbar.cpp:
```diff
@@ -0,0 +1,332 @@
+
+        int imagePos(int pieceIndex) const
+        {
+            return m_bytesPerPixel < 0 ? 0 : (pieceIndex * m_pieceSize + m_pieceSize / 2) / m_bytesPerPixel;
```
@sledgehammer999, I think that with this we avoid float-point arithmetic, but please check.

sledgehammer999 replied:
I don't know why calculations seem off. Example:
**torrent size** = 9 **bytes**
**piece size** = 2 bytes
**image width** = 500 pixels
**bytesPerPixel** = 0.018

Assuming you want piece 5 (last one) then you get:
(5 * 2 + 2 / 2) / 0.0.18 = 11 / 0.018 = 611.111!!!

evsh replied:
the idea behind switching to bytes is that there will be no torrent with size in bytes less than ten times size of screen in pixels.

sledgehammer999 replied:
Sorry I made a mistake. The piece index for the last piece would be 4 not 5. Which gives and answer of 500. But since the image is so much bigger than the torrent size should one byte occupy several pixels?
Shouldn't bytesPerPixel = width / size ?

evsh replied:
In your example `m_bytesPerPixel == 0` (and see my comment above). If you think that probability to encounter a torrent with size less than 1000 bytes is not zero, or you just don't want to leave any chance to it, I will rewrite this with floating-point numbers.

sledgehammer999 replied:
And I think I get wrong results for this example too:
**torrent size** = 10 MiB
**piece size** = 2 MiB
**image width** = 500 pixels
**bytesPerPixel** = 10240 / 500 = 20.48 = 20

Pos for piece 4 (last one):
(4 * 2048 + 2048 / 2) / 20 = 9216 / 20 = 460.80

That means that about 40px are allocated for each piece. So 5 pieces occupy 200px. Where do the rest 300px go?

evsh replied:
I suppose that you meant KiB, not MiB, for the sizes, then bytesPerPixel = 20.

Let's compute pixels per piece: pixelsPerPiece = totalPixels / numberOfPieces = totalPixels / ceil(torrrentSize / pieceSize); In this example numberOfPieces = 5 and pixelsPerPiece = 100. The method computes image positions for the center of the given piece. Exact number is 450, while computation gave 460. The error is 10/450 =  2.2%.

The estimation of the error is: `image.width() / torrentInfo.totalSize() * 100%` = 500 / 10240 *100% = 4.9%;

However, taking a torrent of size 10 MiB we get error value of approx 1920 / (10 * 2^20)*100% =  0.02% and thus a typical error will be 1920*0.02% = 0.384 px.

Should we want precise pointing, we need to compute two image positions: for the beginning of a piece a for the end. Perhaps that would make computations clearer, I don't know. Testing on you showed that I have to either put the approximation used into a comment or rewrite the code using two ends and, perhaps, floating point arithmetic.

sledgehammer999 replied:
Oops you're right. I messed up the units. 

Anyway, I didn't realize you wanted to calculate the position at the center. I think you're explanation is enough and IMO the error rate is acceptable.
More comments on what the algos do are welcome for future reference.

And I suppose torrents smaller than the image are so improbable that we shouldn't worry about them.

evsh replied:
I will add the comment then.

evsh replied:
The function works incorrectly (don't know details yet), but in multi-file torrent it highlights exactly in the first half of the progress bar.

evsh replied:
It was an integer overflow when computing value inside of the braces. Will make `m_pieceSize` of type `qlonglong` too,

evsh commented:
@qbittorrent/qbittorrent-frequent-contributors this PR adds `QPair<int, int> TorrentInfo::piecesForFile(int fileIndex) const` which returns indices of first and last pieces that contain a given file. At the same time, there is `QPair<int, int> TorrentHandle::fileExtremityPieces(int index) const` function with the same purpose. My bad I had not found the latter. However, implementations are different. I believe that `TorrentInfo::piecesForFile` is better, of course. For instance, `TorrentHandle::fileExtremityPieces` incorrectly behaves in corner cases (take files_offset =  n * piece_size - 1 and file_size = n * piece_size). 

So, what should we do? Could someone, please confirm or reject my conclusion (the function in  `TorrentInfo` does not have corner cases, the one in `TorrentHandle` does). And we shall remove one of them. I believe this function belongs to `TorrentInfo` class since it needs only the torrent metadata to work.

glassez commented:
>I believe this function belongs to TorrentInfo class since it needs only the torrent metadata to work.

Ok.
>I believe that  TorrentInfo::piecesForFile  is better, of course. 

At least it has incorrect name.

evsh commented:
> At least it has incorrect name.

Then suggest the correct one, please.

glassez commented on src/base/bittorrent/torrentinfo.cpp:
```diff
@@ -213,24 +220,79 @@ QByteArray TorrentInfo::metadata() const
+    if (!isValid())
+        return qMakePair(-1, -1);
+
+    if ((fileIndex < 0) || (fileIndex >= filesCount()) ) {
```
Remove extra space between last parentheses.

glassez commented:
>Then suggest the correct one, please.

Why not to use the old one? I mean fileExtremityPieces.

glassez commented:
@evsh your method is better (if it's correct of course) since it doesn't use floating point arithmetic.

evsh commented:
> Why not to use the old one? I mean fileExtremityPieces.

Simple and nice :)

> @evsh your method is better (**if it's correct of course**)

That's the question :) I feel like I want to add unit tests to qBt.

glassez commented:
>That's the question :) I feel like I want to add unit tests to qBt.

I think it would be easy to prove it in this case. Enough to do it for the boundary cases.

evsh commented:
I would like to get a confirmation from at least one more person.

glassez commented:
Ok.
`firstOffset / pieceLength` is correct for each correct firstOffset.
But...
`(firstOffset + fileSize) / pieceLength` isn't correct since `(firstOffset + fileSize)` is the next file offset and we should check the last file byte offset instead: `(firstOffset + fileSize - 1)`.

evsh commented:
Bingo! Thank you!

Another question: wouldn't client code become simpler if we return not [first,last] pair, but [first,one after the last) range? Additionally, that will be consistent with STL interfaces and the second number will be equal to the number of pieces, into which the file extends.

glassez commented:
>Another question: wouldn't client code become simpler if we return not [first,last] pair, but [first,one after the last) range? 

What good to us getting "one after the last" piece number?
>Additionally, that will be consistent with STL interfaces and the second number will be equal to the number of pieces, into which the file extends.

"number of pieces, into which the file extends" is not the same as "one after the last file piece"

evsh commented:
Because this is what we used to get from arrays and squences: index of the first element and size (index of the element after the last one).

> "number of pieces, into which the file extends" is not the same as "one after the last file piece"

plus the start index, of course.

glassez commented:
So what value are you going to return? If "number of pieces, into which the file extends" then it requires some more complex algorithm. If "one after the last" then why don't you just cut this method to `fileFirstPieceIndex()` and to call it twice: fileFirstPieceIndex(i) and fileFirstPieceIndex(i + 1)?

Besides, I am confused by what term can be called a pair of different entities (firstPieceIndex, numberOfPieces) or even pair of first and last+1 indexes?

> that will be consistent with STL interfaces
IMO, we should not aspire to be like STL for no additional reason.

evsh commented:
This entity usually called a range. Why is it good? Because having 
```C++
auto range = info.fileRange(fileIndex)
```
one can write enumerations as he is used to:
```C++
for (auto index = range.first; index != range.second; ++index)
for (auto index = range.first; index < range.second; ++index)
```
or use STL algorithms with a range.

glassez commented:
>one can write enumerations

The other can write differently (assuming filePiecesRange returns first and last index):
```c++
auto range = info.filePiecesRange(fileIndex);
for (auto index = range.first; index <= range.second; ++index)
```
Why not?
Besides, we don't have to alter existing code.

glassez commented:
>This entity usually called a range.

I am confused by the fact that your range includes out-of-range value too.
You probably want to show the analogy with the iterators. But it is somewhat different.

evsh commented:
http://www.cs.utexas.edu/~EWD/transcriptions/EWD08xx/EWD831.html

evsh commented:
BTW, with this agreement the code in the function which determines a range is simpler too:  there is no need to subtract 1 from the file size.

glassez commented:
>there is no need to subtract 1 from the file size.

This will work incorrectly in some cases for last file.

evsh commented:
We do integer division thus I think everything would be OK for all mentioned cases.

glassez commented:
>Understood what did you say.

Now you have a simple and reliable algorithm of search the first and last piece. No need to complicate it. Everything else can be done on its basis.

evsh commented:
PR updated: 
1. Instead of `QPair<int, int>` use class `IndexRange<int>`  (added) which allows removing ambiguity of the `QPair` `first` and `second` meanings. Also, explicit return type allowed to simplify the function name which is just `TorrentInfo::filePieces()` now.
2. Added commit which replaces `TorrentHandle::fileExtremityPieces()` with `TorrentInfo::filePieces()` because we do not need the handle for acessing torrent layout.

LordNyriox commented on src/base/bittorrent/torrentinfo.cpp:
```diff
@@ -213,24 +220,79 @@ QByteArray TorrentInfo::metadata() const
 
 QStringList TorrentInfo::filesForPiece(int pieceIndex) const
```
@evsh:  Is `filesForPiece` still used?  I thought you renamed the function?

evsh replied:
I renamed the reverse function (piece range from file index). Probably you were confused by that. And yes, we need both functions.

LordNyriox replied:
@evsh:  O.K. then.  Thanks.

evsh commented on src/base/bittorrent/indexrange.h:
```diff
@@ -0,0 +1,127 @@
+            : m_first {first}
+            , m_last {last}
+        {
+            Q_ASSERT(first <= last);
```
Decided to remove `constexpr` for now. Waiting for C++14.

glassez replied:
Please describe it in short. What changes in C++14 do you wait?

evsh replied:
The relaxed `constexpr` restrictions, of course.

glassez replied:
I thought so. But another question - will it be enough?

evsh replied:
Sorry, I don't understand you. I just want to put an assertion into a constexpr function. One needs C++14 for that.

glassez replied:
I'm not sure these relaxation will be enough to use Q_ASSERT in `constexpr` methods.

evsh commented:
I don't want to change anything more with this PR, unless there will be any suggestions (or errors). @sledgehammer999, consider this for merging, please.

evsh commented:
> I don't want to change anything more with this PR

Broke the promise and uncrustified  `src/gui/properties/propertieswidget.cpp` in the first commit too. This is just to simplify merging in my fork.

sledgehammer999 commented:
>I don't want to change anything more with this PR, unless there will be any suggestions (or errors). @sledgehammer999, consider this for merging, please.

I am going to review this tomorrow.

sledgehammer999 commented on src/base/bittorrent/torrentinfo.h:
```diff
@@ -34,12 +34,15 @@
 class QStringList;
 class QByteArray;
 template<typename T> class QList;
+template<typename T> class QVector;
```
Foward declaring indexrange the same way, doesn't work?

evsh replied:
It works, of course. Decided that the file is so tiny that such optimisation is not significant, while, naturally, this ways it is more easy to use.

sledgehammer999 commented on src/gui/properties/piecesbar.cpp:
```diff
@@ -0,0 +1,318 @@
+    {
+    public:
+        PieceIndexToImagePos(const BitTorrent::TorrentInfo &torrentInfo, const QImage &image)
+            : m_bytesPerPixel {image.width() > 0 ? torrentInfo.totalSize() / image.width() : -1}
```
What if the file size is smaller than the image? Extreme case. Look at my example below.

evsh replied:
It will render incorrectly if `image.width() > torrentInfo.totalSize()`, but I thought that this will never happen. And the relative precision of the computations is `image.width() / torrentInfo.totalSize() * 100%`

evsh replied:
Added a comment which explains the used approximation.

evsh replied:
Just realised that using the `PieceRange` abstraction I can make the interface of this class cleaner and remove one of the approximations (for the piece center).

sledgehammer999 commented on src/gui/properties/piecesbar.cpp:
```diff
@@ -0,0 +1,318 @@
+        stream << "<html><body>";
+        const int imagePos = e->pos().x() - borderWidth;
+        if ((imagePos >=0) && (imagePos < m_image.width())) {
+            PieceIndexToImagePos transform {m_torrent->info(), m_image};
```
Don't use curly braces. Use the parentheses.

evsh replied:
OK, but this is how C++ code will look like in the very near future.

glassez replied:
But we did accept using it!
Let's be consistent in our actions.

@evsh You agreed to this so easily? After you have fought for this opportunity? You surprise me (not to say otherwise).

sledgehammer999 replied:
I went back and read the relevant issue. Ok, I understand why you want it(to narrow the implicit type conversions). I mistakenly thought that it was only to initialize elements of lists (eg vector, map, hash).
You can leave them then. I guess old habits make my eyes hurt by seeing this "unusual/different" invocation.

evsh replied:
Thank you, @glassez

sledgehammer999 commented on src/gui/properties/piecesbar.cpp:
```diff
@@ -0,0 +1,318 @@
+        if ((imagePos >=0) && (imagePos < m_image.width())) {
+            PieceIndexToImagePos transform {m_torrent->info(), m_image};
+            int pieceIndex = transform.pieceIndex(e->pos().x() - borderWidth);
+            QVector<int> files {m_torrent->info().fileIndicesForPiece(pieceIndex)};
```
same here.

sledgehammer999 commented:
Would it be easy for you to undo almost all your curly braces usage? eg on initializer lists and when declaring a variable? It breaks the style the rest of the project uses (and it annoys me reading it).

As for the PR: My only gripe so far is the calculation I commented on. I could be wrong. I am waiting your input.

sledgehammer999 commented on src/gui/properties/piecesbar.cpp:
```diff
@@ -0,0 +1,318 @@
+
+        int pieceIndex(int imagePos) const
+        {
+            return m_bytesPerPixel < 0 ? 0 : (imagePos * m_bytesPerPixel + m_bytesPerPixel / 2) / m_pieceSize;
```
and I don't check this one until the first one is resolved.

evsh commented:
PR updated: approximation is mentioned in a source comment, integer overflow error was corrected.

evsh commented on src/gui/properties/piecesbar.cpp:
```diff
@@ -0,0 +1,332 @@
+                         << "Torrent size =" << torrentInfo.totalSize() << "Image width = " << image.width();
+        }
+
+        ImageRange imagePos(const BitTorrent::TorrentInfo::PieceRange &pieces) const
```
@sledgehammer999, @glassez 
Take a look, please. As to me the new functions is more correct and cleaner.

evsh commented on src/gui/properties/piecesbar.cpp:
```diff
@@ -0,0 +1,332 @@
+
+namespace
+{
+    using ImageRange = BitTorrent::IndexRange<int>;
```
The `IndexRange` type seems to be too general to be a member of the `BitTorrent` namespace?

glassez replied:
Of course. Move it into global namespace and its file in `base/` directly.

glassez commented on src/base/bittorrent/indexrange.h:
```diff
@@ -0,0 +1,133 @@
+namespace BitTorrent
+{
+    // Interval is defined via [first;last]
+    template <class Index>
```
I don't like the use **class** keyword in this context.

evsh replied:
class → typename

glassez commented on src/gui/properties/piecesbar.cpp:
```diff
@@ -0,0 +1,332 @@
+            return m_bytesPerPixel < 0
+                   ? BitTorrent::makeInterval(0, 0)
+                   : BitTorrent::makeInterval<ImageRange::IndexType>(
+                (static_cast<qlonglong>(pieces.first()) * m_torrentInfo.pieceLength()) / m_bytesPerPixel,
```
Some incorrect indentation...

glassez replied:
The use of ternary operator with such cumbersome constructions inside looks horrible.
IMO, in this case it's better (more readable) to use "earlier return" or `if-else`:
```c++
if (m_bytesPerPixel < 0) return makeInterval(0, 0);

return makeInterval<ImageRange::IndexType>(
    (static_cast<qlonglong>(pieces.first()) * m_torrentInfo.pieceLength()) / m_bytesPerPixel,
    (static_cast<qlonglong>(pieces.last()) * m_torrentInfo.pieceLength() + m_torrentInfo.pieceLength(pieces.last() - 1)) / m_bytesPerPixel);
```
```

evsh replied:
Simplified. What's wrong with the indentation?

glassez replied:
>What's wrong with the indentation?

Next indentation level should be more indented than previous one.

evsh replied:
Isn't it the case here?

glassez replied:
>Isn't it the case here?

Your previous code was:
```c++
return m_bytesPerPixel < 0
       ? BitTorrent::makeInterval(0, 0)
       : BitTorrent::makeInterval<ImageRange::IndexType>(
    (static_cast<qlonglong>(pieces.first()) * m_torrentInfo.pieceLength()) / m_bytesPerPixel,
    (static_cast<qlonglong>(pieces.last()) * m_torrentInfo.pieceLength() + m_torrentInfo.pieceLength(pieces.last() - 1)) / m_bytesPerPixel);
```

The last two lines should be more indented than `: BitTorrent::makeInterval<ImageRange::IndexType>(`

evsh replied:
Oh, I see. Thanks!

glassez commented on src/gui/properties/piecesbar.cpp:
```diff
@@ -0,0 +1,334 @@
+
+        ImageRange imagePos(const BitTorrent::TorrentInfo::PieceRange &pieces) const
+        {
+            if  (m_bytesPerPixel < 0)
```
Now here is the extra space...

sledgehammer999 commented:
I think the code is ok for merging. I did a test run and I noticed something different from a previous version I had tested. I don't know if this is intentional or not. When hovering over the progress/availability bars I don't see the relevant piece being highlighted! Or was it the file you were highlighting? In either nothing gets highlighted now.

evsh commented:
> When hovering over the progress/availability bars I don't see the relevant piece being highlighted!

This is a bug, which I can't reproduce. Could you look what happens inside of `PiecesBar::highlightFile(int imagePos)`? In particular, what are the values of `pieceIndex`, `filePieces`, and `imageRange`?

evsh commented:
@sledgehammer999 there was a bug on one of the previous iterations: integer overflow which broke highlighting if position was far than 2GB from the beginning of a torrent. Do you observe something similar?

evsh commented:
and in that case `imageRange` contains negative values.

sledgehammer999 commented on src/base/indexrange.h:
```diff
@@ -0,0 +1,130 @@
+
+    constexpr IndexType last() const
+    {
+        return m_first + m_size - 1;
```
Question: since begin() is the same as first() shouldn't end() be the same as last()?
I am under the impression you offer those for compatibility, right?

evsh replied:
You are right, this id for compatibility, and as such, end() must point to one after the last element.

sledgehammer999 replied:
I suppose you mean last() and end() are both correct as they are now, right?

evsh replied:
Yes.

sledgehammer999 commented:
The issue is that inside `PieceIndexToImagePos::imagePos()` the torrentInfo is invalid so pieceLength is -1. I cannot understand why it is invalid. You are initializing it in the constructor.

On another note:
**2 negative side-effects of curly brace initializations**
1. When using them in the initialization list of a constructor QtCreator gets confused about the rest of the class. For example take class `PieceIndexToImagePos` and go inside imagePos(). Hit enter after `const qlonglong pieceLength = m_torrentInfo.pieceLength();`. The newline gets indented too far to the right.
2. In `QRect newHighlitedRegion {imageRange.first(), 0, imageRange.size(), m_image.height()};` hover over newHighlitedRegion. The tooltip doesn't show the constructor signature.

sledgehammer999 commented:
There's some very wrong here.
If I put in `PiecesBar::highlightFile()` this:
```c++
PieceIndexToImagePos transform {m_torrent->info(), m_image};
qCritical() << m_torrent->info().isValid();
```
it works!!!

Vanilla code(without qCritical()) doesn't work.

evsh commented:
OMG, I thought that `TorrentHandle::info()` returns `const TorrentInfo&`...

evsh commented:
>> There's something very wrong here.

>OMG, I thought that TorrentHandle::info() returns const TorrentInfo&...

Fixed.

evsh commented:
BTW, GCC with -Wextra should warn about this error. Unfortunately, we do not pass neither -Wextra nor other warning options to it.

glassez commented:
>OMG, I thought that TorrentHandle::info() returns const TorrentInfo&... 

@evsh So what was the problem? I was lost in recent code changes...

evsh commented:
@glassez `PieceIndexToImagePos` was keeping const reference to a temporary `TorrentInfo` object, obtained via `TorrentHandle::info()`

sledgehammer999 commented:
I'll be able to check in a couple of hours. In the meantime, please take a look at my  inline comment in `constexpr IndexType last() const`.

PS: Aren't you able to define extra CXXFLAGS before compiling?

evsh commented:
> PS: Aren't you able to define extra CXXFLAGS before compiling?

If I modify cmake files, each git command will ask me to commit or stash those changes.

sledgehammer999 commented:
Thanks for this.

>If I modify cmake files, each git command will ask me to commit or stash those changes.

Not true. It asks about that edited file only when you are rebasing or pulling a branch that modifies that same file. But this is solved easily. You do your necessary changes and stash them. Each time you clean your working dir you just need "git stash apply" to reapply them. The "apply" verb keeps the stash and doesn't delete it. 
This may not work so well if you want to keep multiple stashes. You'll need to remember the index too then.

evsh commented:
Thank you for merging, and I very much thankful to all those who took part in the discussion!

@sledgehammer999, I hope that you ran it one more time and that highlighting work. Does it? 

>  It asks about that edited file only when you are rebasing or pulling a branch that modifies that same file.

And rebasing appears to be the most frequent operation :( Especially to amend a HEAD^n commit.

Chocobo1 commented:
Was just playing this, found a minor issue:
![pic](https://cloud.githubusercontent.com/assets/9395168/15756031/01fb9cd2-2932-11e6-8024-566d29be52e3.png)
This happens when I point to the very start (and end) of the progress bar.

evsh commented:
@Chocobo1: I can't understand what is in your image. What is the dark rectangle? What is supposed to be there?

Chocobo1 commented:
Forgot to say, the box is supposed to show detailed information (holding down shift).

evsh commented:
So the detailed tooltips for the very first piece and the very last ones are empty. Is it correct?

Chocobo1 commented:
Not sure, it only happens on the border.

LordNyriox commented:
@evsh:  It's an empty tooltip box.

evsh commented:
Fixed in #5339
